### ISSUE TITLE
[feature] Add support for the files generation (csv, json, xml)

### ISSUE THREAD

**Issue by NarekDW** [datafaker-net/datafaker#784]
**Is your feature request related to a problem? Please describe.**
What do you think about file generation?
The idea is to generate the fake data directly to the files.
It might be useful for big data developers to generate a big amount of dummy files.

**Describe the solution you'd like**
We could add an API like this:
```java
int generateToFile(String folderPath, Schema<IN, ?> schema, int rowsPerFile, int numberOfFiles, boolean overwrite);
```
which will generate files with fake data, for example csv files, like this:
<img width="171" alt="image" src="https://user-images.githubusercontent.com/24260474/233805825-651196a2-86ba-4ed7-ba89-06d6f2ee3858.png">

**Describe alternatives you've considered**
Alternative is to generate fake data as String (in the memory) and then write it to the files manually.

**Comment by kingthorin** [comment]
https://github.com/datafaker-net/datafaker/tree/main/src/main/java/net/datafaker/transformations

**Comment by NarekDW** [comment]
> * https://github.com/datafaker-net/datafaker/tree/main/src/main/java/net/datafaker/transformations
> * https://github.com/datafaker-net/datafaker#formats

Yep, that API allows to generate file formats in memory only (it returns String representation of the file). But the idea is to write the data directly to the files on OS.

**Comment by kingthorin** [comment]
Writing to disk isn’t exactly heavy lifting. I mean I guess we could do that for users but it seems kinda outside the purview of the library.

A couple calls to nio and you’re done.

**Comment by bodiam** [comment]
I agree with @kingthorin here. 

Before we know it we have files, streams, writers, etc. In this case, you could use Files.write or something similar, which would be just as easy.

**Comment by bodiam** [comment]
Btw, what I wouldn't mind btw, is "datafaker-cli", which could be a tool to generate fake data, and which could directly write to file in csv/xml/ etc format.

**Comment by snuyanzin** [comment]
I agree
moreover nowadays it doesn't cover all the cases:
e.g. what if I want to write to binary files or to files with some compression e.g. snappy, bzip, etc...
in my opinion it could be a kind of plugin to datafaker however not in core datafaker since besides files there could be similar approach for jdbc, kafka, pulsar and etc... 

>"datafaker-cli"

since we are on jdk17+ there is a jshell =) and some doc in readme how to use it

**Comment by bodiam** [comment]
jshell is nice, but that would require still some hassle to get it working. Something like:

`./datafaker-cli generate --format=csv fields=name.firstName,name.lastName,oscar.quotes --amount=100000 --output=output.csv` 

to me seems easier, and can be scripted etc.

PS: do we have documentation on jshell? I can't find it.

**Comment by snuyanzin** [comment]
>PS: do we have documentation on jshell? I can't find it.

in README `get started`  https://github.com/datafaker-net/datafaker#get-started

there is first java code, then kotlin, then JShell

**Comment by bodiam** [comment]
Ah, thanks. I only checked our mkdocs documentation. Still, I don't think jshell is an alternative to a cli

**Comment by snuyanzin** [comment]
depends on use case
in case of jshell it could done via 
1. From project root folder `jshell --class-path $(ls -d target/*.jar | tr '\n' ':')`
2. type `/edit` to enter in multiline edit mode
3. insert 
```java
import net.datafaker.Faker;
import net.datafaker.providers.base.Name;
import net.datafaker.transformations.Schema;
import net.datafaker.transformations.CsvTransformer;
import static net.datafaker.transformations.Field.field;
var faker = new Faker();

Schema<Name, String> fromScratch =
    Schema.of(field("firstName", () -> faker.name().firstName()),
        field("lastname", () -> faker.name().lastName()));
CsvTransformer<Name> transformer =
    CsvTransformer.<Name>builder().header(false).separator(",").build();
try (BufferedWriter fw = Files.newBufferedWriter(Path.of("output.csv"))) {
    fw.write(transformer.generate(fromScratch, 2));
} catch (IOException e) {
    throw new RuntimeException(e);
}
```
4.  `Ctrl-X`
it will save output in `output.csv`

**Comment by bodiam** [comment]
I think your definition of user friendly strongly deviates from mine.

**Comment by NarekDW** [comment]
> I agree moreover nowadays it doesn't cover all the cases: e.g. what if I want to write to binary files or to files with some compression e.g. snappy, bzip, etc... in my opinion it could be a kind of plugin to datafaker however not in core datafaker since besides files there could be similar approach for jdbc, kafka, pulsar and etc...
> 
> > "datafaker-cli"
> 
> since we are on jdk17+ there is a jshell =) and some doc in readme how to use it

That's exactly what I was thinking about... That would be nice to have such an ability (plugin) to generate fake data directly to databases or distributed systems, like kafka, hdfs and so on, in different formats and compressions...

**Comment by NarekDW** [comment]
> depends on use case in case of jshell it could done via
> 
> 1. From project root folder `jshell --class-path $(ls -d target/*.jar | tr '\n' ':')`
> 2. type `/edit` to enter in multiline edit mode
> 3. insert
> 
> ```java
> import net.datafaker.Faker;
> import net.datafaker.providers.base.Name;
> import net.datafaker.transformations.Schema;
> import net.datafaker.transformations.CsvTransformer;
> import static net.datafaker.transformations.Field.field;
> var faker = new Faker();
> 
> Schema<Name, String> fromScratch =
>     Schema.of(field("firstName", () -> faker.name().firstName()),
>         field("lastname", () -> faker.name().lastName()));
> CsvTransformer<Name> transformer =
>     CsvTransformer.<Name>builder().header(false).separator(",").build();
> try (BufferedWriter fw = Files.newBufferedWriter(Path.of("output.csv"))) {
>     fw.write(transformer.generate(fromScratch, 2));
> } catch (IOException e) {
>     throw new RuntimeException(e);
> }
> ```
> 
> 4. `Ctrl-X`
>    it will save output in `output.csv`

This looks a bit complicated. 
What about using `jline` for example and introduce our custom query language for such kind of operations?))

**Comment by bodiam** [comment]
Just an FYI (thanks @panilya for letting me know): https://github.com/bernhardhuber/datafaker-cli

**Comment by kingthorin** [comment]
So this can be closed?
Or, an admin can move it to the new repo?

**Comment by bodiam** [comment]
@kingthorin ? Which new repo?

**Comment by kingthorin** [comment]
Oh sorry I missed that was third party 🤦‍♂️

**Comment by bodiam** [comment]
Ahh, got it! Sorry for misunderstanding. 

Yeah, it's third party, but something like this might be nice to incorporate into the main repo, probably as a new module.

**Comment by snuyanzin** [comment]
>What about using jline for example and introduce our custom query language for such kind of operations?))

why do we need another custom language if there are already expressions?

>That's exactly what I was thinking about... That would be nice to have such an ability (plugin) to generate fake data directly to databases or distributed systems, like kafka, hdfs and so on, in different formats and compressions...

I do not understand this.
Why from one side we are talking about generation of data in specific formats and from another side we are talking about jline which is cli i.e. something to communicate with user?

How could jline help here? Why just input arg are not enough?
e.g.
```
./datafaker-whatever-you-like-plugin generate --format=csv fields=name.firstName,name.lastName,oscar.quotes --amount=100000 --output=output.csv
```

**Comment by NarekDW** [comment]
Yep, you are right no need for jline, as we don't need an interactive shell for that purpose, I was thinking that it might be more convenient for users.

In case of args:
`./datafaker-whatever-you-like-plugin generate --format=csv fields=name.firstName,name.lastName,oscar.quotes --amount=100000 --output=output.csv`

What about generation of complex field type (like struct, array, etc...) ? 
What about nullability, min and max values, column name, data type, etc ? 
Would it be possible to configure such properties?

**Comment by snuyanzin** [comment]
there are at least 2 ways and both could be implemented: pass it via flags/args and put it in conf file and pass conf file.

Complex types with names could be implemented with use of schemas
do you see any issue here?

**Comment by NarekDW** [comment]
No, I don't, but I can't imagine for now, how it could be configured (conf file for example, should it be a json file with fixed schema and props?), I'm afraid it might become too complicated for users.

**Comment by snuyanzin** [comment]
any proposal to simplify it welcome

**Comment by NarekDW** [comment]
Okay lets do it in that way.  It is fine to use bash script to generate files, but it also will be nice to support files schemas in separate files (for generating files with large or complicated schemas).
For example:
```bash
./datafaker-whatever-you-like-plugin generate 
   format=CSV|JSON|XML|PARQUET|AVRO... 
   schemas=[path_to_schema_1, path_to_schema2, ...]
   rowsPerFile=100000 
   numberOfIles=10
   ... (and so on)
```

And we can use separate json or yaml files to describe file schema:
```yaml
field_1:
    type: string
    nullable: true
    maxLength: 10
    minLength: 2
    generators: [name.firstName, name.lastName, ...]
field_2:
    type: array
    nullable: false
    maxLength: 100
    generators: [name.firstName, name.lastName, ...]
field_3:
    type: struct
    nullable: false
    fields:
        inner_field1:
             type: int
             nullable: true
             generators: [name.firstName, name.lastName, ...]
             ....
        inner_field2:
...

```

what do you think?

**Comment by snuyanzin** [comment]
Can you elaborate what `maxLength`/`minLength` means in case of type `string`? Is it a length of the field or what?
I guess it's better to specify not `nullable` as boolean but `nullRate` as double. In that case we can say that it is not only nullable but how often it could be nullable.

**Comment by NarekDW** [comment]
Yep, `maxLength/minLength` - it is meant to be the field length, but it might be redundant for string type, (it was just a random example) probably it will make sense for `array` types only.
Regarding `nullRate` - I agree, sounds good.

**Comment by bodiam** [comment]
I like the Schema idea, but I already built something like it: http://www.schematix.net/docs/modules/data-generation (it's based on Datafaker). The idea was to use JsonSchemas (since they're pretty standard), with a few extension to generate data, validate CSVs, etc. Is that something similar to what you have in mind @NarekDW ?

**Comment by bodiam** [comment]
This is a bit more extended example, I think similar to yours @NarekDW:

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://schema.schematix.net/data-gen/simple_types.json",
  "type": "object",
  "properties": {
    "age": {
      "x-schematix" : {
        "generation-strategy": "natural"
      },
      "type": "string"
    },
    "name": {
      "x-schematix" : {
        "generation-strategy": "schema"
      },
      "type": "string",
      "enum": ["duck", "goose"]
    },
    "breed": {
      "type": "string",
      "x-schematix" : {
        "generation-strategy": "example",
        "scale" : 1
      },
      "pattern": "[a-z]+",
      "examples": [
        "bulldog",
        "pug",
        "poodle"
      ]
    }
  }
}
```

**Comment by snuyanzin** [comment]
@bodiam is it open sourced?
i failed to find any source code for that only this https://github.com/schematix-net

**Comment by NarekDW** [comment]
@bodiam Yep it is what I meant, but it looks a bit complicated at first glance) 
Do you want to make it as a separate project/repository?

**Comment by snuyanzin** [comment]
I would make it as a separate repo
one of the reasons is that it can be done independent on the under the hood engine: it could be datafaker core or any other lib

**Comment by NarekDW** [comment]
> I would make it as a separate repo one of the reasons is that it can be done independent on under the hood engine: it could be datafaker core or any other lib
> 
> could be under [datafaker-net](https://github.com/datafaker-net)

Sounds good to me

**Comment by NarekDW** [comment]
I have couple of questions regarding that schema:
1. why do we need this fields?
```json
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://schema.schematix.net/data-gen/simple_types.json",
  "type": "object"
```
2. As I understood generation-strategy example - means that it will pick one of the `examples`'s field values, why do we need `pattern` and `scale` fields here?
```json
"x-schematix" : {
        "generation-strategy": "example",
        "scale" : 1
      },
      "pattern": "[a-z]+",
      "examples": [
        "bulldog",
        "pug",
        "poodle"
      ]
```

**Comment by bodiam** [comment]
> 1. why do we need this fields?

Because it's part of the Json schema specification.

> 2. why do we need pattern and scale fields here

Good questions. We don't, I think that's a mistake in my example, scale really doesn't make sense :)

**Comment by snuyanzin** [comment]
a couple more ideas:
both formats and outputs could implement SPI
this will allow to connect new formats/output options without rebuilding the tool, just by adding a corresponding jar with implementation

**Comment by bodiam** [comment]
> @bodiam is it open sourced? i failed to find any source code for that only this https://github.com/schematix-net

It's not really open source. I could make it open source, cause I don't do much with it really. I am open to the idea.

**Comment by bodiam** [comment]
> @bodiam Yep it is what I meant, but it looks a bit complicated at first glance) Do you want to make it as a separate project/repository?

It looks a bit complicated indeed, especially at first glance, and it didn't start this complex, but you add more features, and you want a bit more flexibility, and you sort of end up like this. It doesn't have to be a JSON schema, but you want something like a schema, and this could be defined in any way you want (YML, XML, plain text, whatever), but in the end, you want some flexibility in your data generation. The idea was that JSON schemas are indeed a bit complex, but there could be an editor for it to make things simpler, plus a repository to store these things, especially for more industry standard things, like ICD10AM for example. That way a schema needs to be created only once, and you can reuse them.

**Comment by snuyanzin** [comment]
From one side I would tend to agree that JSON Schema is complicated, from another it's widely used...

So the suggestion from my side is to think of input format parser also pluggable with requirement that this pluggable format should parse it's own config and translate the result into datafaker's Schema. This will allow to have both YML and JSON Schema input formats and may be something else including custom.


I would suggest to have something like POC under https://github.com/datafaker-net and then let's see.
Any objections @bodiam , @NarekDW ?
Also @RVRhub would like to participate IIUC.

**Comment by NarekDW** [comment]
> From one side I would tend to agree that JSON Schema is complicated, from another it's widely used...
> 
> So the suggestion from my side is to think of input format parser also pluggable with requirement that this pluggable format should parse it's own config and translate the result into datafaker's Schema. This will allow to have both YML and JSON Schema input formats and may be something else including custom.
> 
> I would suggest to have something like POC under https://github.com/datafaker-net and then let's see. Any objections @bodiam , @NarekDW ? Also @RVRhub would like to participate IIUC.

No objections, let’s start, I'll be glad to help with development :)

**Comment by bodiam** [comment]
No objections here either!

**Comment by snuyanzin** [comment]
there is a POC https://github.com/snuyanzin/datafaker-gen
seems i don't have enough grants to create it under datafaker-net

**Comment by snuyanzin** [comment]
since there is working solution at already mentioned  https://github.com/snuyanzin/datafaker-gen
which also contains a doc how to use
i would consider this as not an issue any more

**Comment by snuyanzin** [comment]
Also now under datafaker-net
https://github.com/datafaker-net/datafaker-gen
